File: libs/python/vidi.py
import httplib, urllib
import simplejson as json
import random


class Vidi(object):
    def __init__(self, apikey, server="api.vidi.zeitin.com"):
        self.apikey = apikey
        self.id = apikey
        self.rest_host = server

    def get_rooms(self):
        roomids = self.request('rooms', 'GET')
        return [Room(self, roomid) for roomid in roomids]

    def get_room(self, roomid):
        return Room(self, roomid)

    def create_room(self):
        roomid = self.request('rooms/create', 'POST')[0]
        return Room(self, roomid)

    def get_property(self, key, access='private'):
        value = self.request('properties/get', 'GET', {
            'where': 'apikey',
            'id': self.id,
            'key': key,
            'access': access,
        })
        return value and value[0]

    def set_property(self, key, value, access='private'):
        result = self.request('properties/set', 'POST', {
            'where': 'apikey',
            'id': self.id,
            'key': key,
            'value': value,
            'access': access,
        })[0]
        return result

    def get_desktop(self, desktopid):
        return Desktop(self, desktopid)

    def close(self):
        self.request('session/destroy', 'DELETE')

    def get_init_js(self, room, client, callback=False, debug=False):
        if isinstance(client, Client) == False:
            raise VidiError('Wrong Client object')
        if isinstance(room, Room) == False:
            raise VidiError('Wrong Room object')

        params = {
            'clientid': client.id,
            'roomid': room.id,
            'debug': debug,
        }

        if callback:
            params.update({'callback': callback})

        return """
<script type="text/javascript" src="%(vidi_js)s"></script>
<script type="text/javascript">
vidi.initialize(%(params)s);
</script>
        """ % {
            'vidi_js': 'http://' + self.rest_host + '/vidi/static/vidi.js',
            'params': json.dumps(params),
        }

    def create_screen(self, **kwargs):
        #TODO: validate params
        divid = 'vidi_screen_' + str(random.randint(1, 10000))
        self.oo = kwargs
        if self.oo.has_key('input'):
            if isinstance(self.oo['input'], Input) == False:
                raise VidiError('Wrong Input object')
            self.oo['inputid'] = self.oo['input'].id
            del self.oo['input']
        if self.oo.has_key('output'):
            if isinstance(self.oo['output'], Output) == False:
                raise VidiError('Wrong Output object')
            self.oo['outputid'] = self.oo['output'].id
            del self.oo['output']
        if self.oo.has_key('divid'):
            html = """
<script type="text/javascript">
vidi.createScreen(%(params)s);
</script>
            """ % {
                'params': json.dumps(self.oo),
            }
        else:
            self.oo['divid'] = divid
            html = """
<div id="%(divid)s"></div>
<script type="text/javascript">
    vidi.createScreen(%(params)s);
</script>
            """ % {
                'params': json.dumps(self.oo),
                'divid': divid,
            }
        self.__dict__.update(self.oo)
        return html

    def request(self, url, method, parameters={}):
        path = '/vidi/rest/' + url
        parameters['apikey'] = self.apikey
        params = urllib.urlencode(parameters)
        conn = httplib.HTTPConnection(self.rest_host)
        try:
            if method == 'POST':
                headers = {
                    "Content-type": "application/x-www-form-urlencoded",
                }
                conn.request(method, path, params, headers)
            else:
                headers = {}
                conn.request(method, path + '?' + params, '', headers)
        except Exception, e:
            raise VidiError(e)
        response = conn.getresponse()
        if response.status == 200:
            data = json.loads(response.read())
            try:
                if data.has_key('error') and data['error']:
                    raise VidiError(data['error_desc'])
            except AttributeError:
                pass
            conn.close()
            return data
        else:
            raise VidiError(response.read())

    # GET style rest api
    def __getattr__(self, name):
        def wrapper(**kwargs):
            return self.request(name, 'GET', kwargs)
        return wrapper

    def __repr__(self):
        return '<Vidi Session Object (apikey: %s, server: %s)>' % (self.apikey, self.rest_host)

class Room(object):
    def __init__(self, vidi, roomid):
        self.vidi = vidi
        self.roomid = roomid
        self.id = roomid

    def get_clients(self):
        clientids = self.vidi.request('clients', 'GET', {
            'roomid': self.roomid,
        })
        return [Client(self.vidi, self, clientid) for clientid in clientids]

    def get_client(self, clientid):
        return Client(self.vidi, self, clientid)

    def get_bindings(self):
        bindingids = self.vidi.request('bindings', 'GET', {
            'roomid': self.roomid,
        })
        return [Binding(self.vidi, self, None, None, bindingid) for bindingid in bindingids]

    def get_binding(self, bindingid):
        return Binding(self.vidi, self, None, None, bindingid)

    def create_client(self):
        clientid = self.vidi.request('clients/create', 'POST', {
            'roomid': self.roomid,
        })[0]
        return Client(self.vidi, self, clientid)

    def bind(self, input, output):
        bindingid = self.vidi.request('bindings/create', 'POST', {
            'inputid': input.inputid,
            'outputid': output.outputid,
        })[0]
        return Binding(self.vidi, self, input, output, bindingid)

    def send_message(self, message):
        self.vidi.request('rooms/send_message', 'POST', {
            'roomid': self.roomid,
            'message': message,
        })

    def close(self):
        self.vidi.request('rooms/destroy', 'DELETE', {
            'roomid': self.roomid,
        })

    def get_property(self, key, access='private'):
        value = self.vidi.request('properties/get', 'GET', {
            'where': 'roomid',
            'id': self.roomid,
            'key': key,
            'access': access,
        })
        return value and value[0]

    def set_property(self, key, value, access='private'):
        result = self.vidi.request('properties/set', 'POST', {
            'where': 'roomid',
            'id': self.roomid,
            'key': key,
            'value': value,
            'access': access,
        })[0]
        return result

    def __repr__(self):
        return '<Vidi Room Object (roomid: %s)>' % self.roomid

class Client(object):
    def __init__(self, vidi, room, clientid):
        self.vidi = vidi
        self.room = room
        self.clientid = clientid
        self.id = clientid

    def get_inputs(self):
        inputids = self.vidi.request('inputs', 'GET', {
            'clientid': self.clientid,
        })
        return [Input(self.vidi, self.room, self, inputid) for inputid in inputids]

    def get_input(self, inputid):
        return Input(self.vidi, self.room, self, inputid)

    def get_outputs(self):
        outputids = self.vidi.request('outputs', 'GET', {
            'clientid': self.clientid,
        })
        return [Output(self.vidi, self.room, self, outputid) for outputid in outputids]

    def get_output(self, outputid):
        return Output(self.vidi, self.room, self, outputid)

    def create_input(self):
        inputid = self.vidi.request('inputs/create', 'POST', {
            'clientid': self.clientid,
        })[0]
        return Input(self.vidi, self.room, self, inputid)

    def create_output(self):
        outputid = self.vidi.request('outputs/create', 'POST', {
            'clientid': self.clientid,
        })[0]
        return Output(self.vidi, self.room, self, outputid)

    def send_message(self, message):
        self.vidi.request('clients/send_message', 'POST', {
            'clientid': self.clientid,
            'message': message,
        })

    def close(self):
        self.vidi.request('clients/destroy', 'DELETE', {
            'clientid': self.clientid,
        })

    def get_property(self, key, access='private'):
        value = self.vidi.request('properties/get', 'GET', {
            'where': 'clientid',
            'id': self.clientid,
            'key': key,
            'access': access,
        })
        return value and value[0]

    def set_property(self, key, value, access='private'):
        result = self.vidi.request('properties/set', 'POST', {
            'where': 'clientid',
            'id': self.clientid,
            'key': key,
            'value': value,
            'access': access,
        })[0]
        return result

    def __repr__(self):
        return '<Vidi Client Object (clientid: %s)>' % self.clientid

class Input(object):
    def __init__(self, vidi, room, client, inputid):
        self.vidi = vidi
        self.room = room
        self.client = client
        self.inputid = inputid
        self.id = inputid

    def play_video(self, videofile):
        self.vidi.request('inputs/play_video', 'PUT', {
            'inputid': self.inputid,
            'videofile': videofile,
        })

    def close(self):
        self.vidi.request('inputs/destroy', 'DELETE', {
            'inputid': self.inputid,
        })

    def play_video(self, videofile):
        self.vidi.request('inputs/play_video', 'GET', {
            'inputid': self.inputid,
            'videofile': videofile,
        })

    def is_active(self):
        value = self.vidi.request('inputs/isactive', 'GET', {
            'inputid': self.inputid,
        })
        return value and value[0]

    def get_property(self, key, access='private'):
        value = self.vidi.request('properties/get', 'GET', {
            'where': 'inputid',
            'id': self.inputid,
            'key': key,
            'access': access,
        })
        return value and value[0]

    def set_property(self, key, value, access='private'):
        result = self.vidi.request('properties/set', 'POST', {
            'where': 'inputid',
            'id': self.inputid,
            'key': key,
            'value': value,
            'access': access,
        })[0]
        return result

    def __repr__(self):
        return '<Vidi Input Object (inputid: %s)>' % self.inputid

class Output(object):
    def __init__(self, vidi, room, client, outputid):
        self.vidi = vidi
        self.room = room
        self.client = client
        self.outputid = outputid
        self.id = outputid

    def close(self):
        self.vidi.request('outputs/destroy', 'DELETE', {
            'outputid': self.outputid,
        })

    def is_active(self):
        value = self.vidi.request('outputs/isactive', 'GET', {
            'outputid': self.outputid,
        })
        return value and value[0]

    def get_property(self, key, access='private'):
        value = self.vidi.request('properties/get', 'GET', {
            'where': 'outputid',
            'id': self.outputid,
            'key': key,
            'access': access,
        })
        return value and value[0]

    def set_property(self, key, value, access='private'):
        result = self.vidi.request('properties/set', 'POST', {
            'where': 'outputid',
            'id': self.outputid,
            'key': key,
            'value': value,
            'access': access,
        })[0]
        return result

    def __repr__(self):
        return '<Vidi Output Object (outputid: %s)>' % self.outputid

class Binding(object):
    def __init__(self, vidi, room, input, output, bindingid):
        self.vidi = vidi
        self.room = room
        self.input = input
        self.output = output
        self.bindingid = bindingid
        self.id = bindingid

    def get_input(self):
        if self.input is None:
            self._populate_io()
        return self.input

    def get_output(self):
        if self.output is None:
            self._populate_io()
        return self.output

    def close(self):
        self.vidi.request('bindings/destroy', 'DELETE', {
            'bindingid': self.bindingid,
        })

    def _populate_io(self):
        io = self.vidi.request('bindings/io', 'GET', {
            'bindingid': self.bindingid,
        })
        self.input = Input(self.vidi, self.room, None, io[0])
        self.output = Output(self.vidi, self.room, None, io[1])

    def is_active(self):
        value = self.vidi.request('bindings/isactive', 'GET', {
            'inputid': self.bindingid,
        })
        return value and value[0]

    def get_property(self, key, access='private'):
        value = self.vidi.request('properties/get', 'GET', {
            'where': 'bindingid',
            'id': self.bindingid,
            'key': key,
            'access': access,
        })
        return value and value[0]

    def set_property(self, key, value, access='private'):
        result = self.vidi.request('properties/set', 'POST', {
            'where': 'bindingid',
            'id': self.bindingid,
            'key': key,
            'value': value,
            'access': access,
        })[0]
        return result

    def __repr__(self):
        return '<Vidi Binding Object (bindingid: %s)>' % self.bindingid

class Desktop(object):
    def __init__(self, vidi, desktopid):
        self.vidi = vidi
        self.desktopid = desktopid
        self.id = desktopid

    def notify(self, message):
        self.vidi.request('desktop/notify', 'POST', {
            'desktopid': self.desktopid,
            'message': message,
        })

    def ring(self, count='1'):
        self.vidi.request('desktop/ring', 'POST', {
            'desktopid': self.desktopid,
            'count': count,
        })

    def __repr__(self):
        return '<Vidi Desktop Object (desktopid: %s)>' % self.desktopid

class VidiError(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)

File: libs/java/src/com/zeitin/vidi/Vidi.java
package com.zeitin.vidi;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.JSONValue;

public class Vidi {

	public static final String PRIVATE = "private";
	public static final String PUBLIC = "public";
	private String address;
	private String apikey;
	private String REST_HOST = "http://api.vidi.zeitin.com";
	private String REST_URL = "/vidi/rest";

	public Vidi() {
		this.address = REST_HOST + REST_URL;
		this.apikey = "a1d386c64bceaf644a8e7093d523eafb4aaa3923";
	}

	public String getApikey() {
		return this.apikey;
	}

	public ArrayList<Room> getRooms() {
		Map<String, String> parameters = new HashMap<String, String>();
		String url = address + "/rooms";
		parameters.put("apikey", apikey);
		String response = request(url, "GET", parameters);
		String roomIds[] = response.split(" ");
		ArrayList<Room> rooms = new ArrayList<Room>();
		for (int i = 0; i < roomIds.length; i++) {
			rooms.add(new Room(this, roomIds[i]));
		}
		return rooms;
	}

	public Room getRoom(String roomid) {
		return new Room(this, roomid);
	}

	public Room createRoom() {
		Map<String, String> parameters = new HashMap<String, String>();
		String url = address + "/rooms/create";
		parameters.put("apikey", apikey);
		String roomId = request(url, "POST", parameters);
		return new Room(this, roomId);
	}

	public Desktop getDesktop(String desktopid) {
		return new Desktop(this, desktopid);
	}

	public void close() {
		Map<String, String> parameters = new HashMap<String, String>();
		String url = this.address + "/session/destroy";
		parameters.put("apikey", this.apikey);
		String response = request(url, "DELETE", parameters);
		System.out.println("vidi close response: " + response);
	}

	public String getAddress() {
		return address;
	}

	public void setAddress(String address) {
		this.address = address;
	}

	public String request(String requestUrl, String method,
			Map<String, String> parameters) {
		String json = "";
		String data = "";
		Iterator<String> iterator = parameters.keySet().iterator();
		String next;
		if (method.equals("GET") || method.equals("DELETE")) {
			try {
				next = iterator.next();
				data = next + "=" + parameters.get(next);
				while (iterator.hasNext()) {
					next = iterator.next();
					data += "&" + next + "=" + parameters.get(next);
				}
				URL url = new URL(requestUrl + "?" + data);
				HttpURLConnection conn = (HttpURLConnection) url
				.openConnection();
				String encoding = URLEncoder.encode(data, "UTF-8");
				conn.setRequestProperty("Authorization", "Basic " + encoding);
				conn.setRequestMethod(method);
				conn.connect();
				InputStream in = conn.getInputStream();
				BufferedReader reader = new BufferedReader(
						new InputStreamReader(in));
				json = reader.readLine();
				conn.disconnect();
			} catch (IOException ex) {
				ex.printStackTrace();
				System.out.println("made it here");
			}
		} else if (method.equals("POST")) {
			try {
				next = iterator.next();
				data = URLEncoder.encode(next, "UTF-8") + "="
				+ URLEncoder.encode(parameters.get(next), "UTF-8");
				while (iterator.hasNext()) {
					next = iterator.next();
					data += "&" + URLEncoder.encode(next, "UTF-8") + "="
					+ URLEncoder.encode(parameters.get(next), "UTF-8");
				}
				// Send data
				URL url = new URL(requestUrl);
				URLConnection conn = url.openConnection();
				conn.setDoOutput(true);
				OutputStreamWriter wr = new OutputStreamWriter(conn
						.getOutputStream());
				wr.write(data);
				wr.flush();
				// Get the response
				BufferedReader rd = new BufferedReader(new InputStreamReader(
						conn.getInputStream()));
				json = rd.readLine();
				wr.close();
				rd.close();

			} catch (IOException ex) {
				ex.printStackTrace();
				System.out.println("made it here");
			}
		}
		Object obj = JSONValue.parse(json);
		JSONArray array = (JSONArray) obj;
		String response = array.get(0).toString();
		for (int i = 1; i < array.size(); i++) {
			response += " " + array.get(i).toString();
		}
		return response;
	}

	// TODO make access parameter optional
	public String _getProperty(String where, String id, String key,
			String access) {
		Map<String, String> parameters = new HashMap<String, String>();
		String url = address + "/properties/get";
		parameters.put("apikey", this.apikey);
		parameters.put("where", where);
		parameters.put("id", id);
		parameters.put("key", key);
		parameters.put("?access", access);
		String value = request(url, "GET", parameters);
		return value;
	}

	public void _setProperty(String where, String id, String key, String value,
			String access) {
		Map<String, String> parameters = new HashMap<String, String>();
		String url = address + "/properties/set";
		parameters.put("apikey", this.apikey);
		parameters.put("where", where);
		parameters.put("id", id);
		parameters.put("key", key);
		parameters.put("value", value);
		parameters.put("?access", access);
		String response = request(url, "POST", parameters);
		System.out.println(response);
	}

	public ArrayList<String> _listProperties(String where, String id,
			String access) {
		Map<String, String> parameters = new HashMap<String, String>();
		String url = address + "/properties";
		parameters.put("apikey", this.apikey);
		parameters.put("where", where);
		parameters.put("id", id);
		parameters.put("?access", access);
		String response = request(url, "GET", parameters);
		String keys[] = response.split(" ");
		ArrayList<String> properties = new ArrayList<String>();
		for (int i = 0; i < keys.length; i++) {
			properties.add(keys[i]);
		}
		return properties;
	}

	public String getProperty(String key, String access) {
		return _getProperty("apikey", this.apikey, key, access);
	}

	public void setProperty(String key, String value, String access) {
		_setProperty("apikey", this.apikey, key, value, access);
	}

	public ArrayList<String> listProperties(String access) {
		return _listProperties("apikey", this.apikey, access);
	}

	public String getInitJS(Room room, Client client, String callback,
			boolean debug) {
		HashMap<String, String> params = new HashMap<String, String>();
		params.put("clientid", client.getId());
		params.put("roomid", room.getId());
		params.put("debug", String.valueOf(debug));
		if(callback != null) {
			params.put("callback", callback);
		}
		String format = "<script type=\"text/javascript\" src=\"" + REST_HOST
		+ "/vidi/static/vidi.js\"></script>";
		format += "\n<script type=\"text/javascript\">";
		format += "\nvidi.initialize(";
		Object obj = JSONObject.toJSONString(params);
		format += obj.toString() + ");";
		format += "\n</script>";
		return format;
	}

	public String getInitJS(Room room, Client client, String callback) {
		return getInitJS(room, client, callback, false);
	}

	public String getInitJS(Room room, Client client, boolean debug) {
		return getInitJS(room, client, null, debug);
	}

	public String getInitJS(Room room, Client client) {
		return getInitJS(room, client, null, false);
	}

	public String createScreen(Object... args) {
		String divid = "vidi_screen_" + (int) (Math.random() * 10000 + 1);
		HashMap<String, String> params = new HashMap<String, String>();
		if(contain("Input", args) != -1) {	
			Input input = (Input) args[contain("Input", args)];
			params.put("inputid", input.getId());
		}
		if(contain("Output", args) != -1) {	
			Output output = (Output) args[contain("Output", args)];
			params.put("outputid", output.getId());
		}
		String html;
		if(contain("String", args) != -1) {
			html = "<script type=\"text/javascript\">";
			html += "\nvidi.createScreen(";
			html += JSONObject.toJSONString(params).toString() + ");";;
			html += "\n</script>";
		} else {
			html = "<div id=\"" + divid + "\"></div>";
			html += "<script type=\"text/javascript\">";
			html += "\nvidi.createScreen(";
			html += JSONObject.toJSONString(params).toString() + ");";
			html += "\n</script>";
		}
		return html;
	}

	private int contain(String key, Object args[]) {
		for(int i = 0; i < args.length; i++) {
			if(args[i].getClass().getName().equals(key)) {
				return i;
			}
		}
		return -1;
	}
}
